hivemind/utils/mpfuture.py

```diff
@@ -111,91 +101,169 @@ def _set_event_threadsafe(self):
+                        state_updated_event.set()
+
+                elif future is None:
+                    logger.debug(
+                        f"Received {msg_type} for MPFuture uid={uid}, but future is already done or destroyed"
+                    )
+                elif msg_type == MessageType.RESULT:
+                    future.set_result(payload)
+                elif msg_type == MessageType.EXCEPTION:
+                    future.set_exception(payload)
+                elif msg_type == MessageType.RUNNING:
+                    try:
+                        future.set_running_or_notify_cancel()
+                    except (InvalidStateError, RuntimeError) as e:
+                        logger.debug(f"Could not set MPFuture (uid={uid}) to running due to {e}", exc_info=True)
+                elif msg_type == MessageType.CANCEL:
+                    future.cancel()
                 else:
-                    raise RuntimeError(f"Received unexpected update type {update_type}")
-            except (BrokenPipeError, EOFError):
+                    raise RuntimeError(f"Received unexpected update type {msg_type}")
+
+                if future is None or future.done():
+                    cls._active_futures.pop(uid, None)
+
+            except (BrokenPipeError, EOFError, ConnectionError):
                 logger.debug(f"Update pipe was was shut down unexpectedly (pid={pid})")
             except Exception as e:
                 logger.exception(f"Could not retrieve update: caught {repr(e)} (pid={pid})")
 
-    def _send_update(self, update_type: UpdateType, payload: Any = None):
+    def _send_update(self, update_type: MessageType, payload: Any = None):
         """This method sends result, exception or cancel to the MPFuture origin."""
-        with MPFuture._update_lock if self._use_lock else nullcontext():
-            self._sender_pipe.send((self._uid, update_type, payload))
+        try:
+            with MPFuture._update_lock if self._use_lock else nullcontext():
+                self._sender_pipe.send((self._uid, update_type, payload))
+        except (ConnectionError, BrokenPipeError, EOFError) as e:
+            logger.debug(f"No updates were sent: pipe to origin process is no longer operational ({e}).", exc_info=True)
+
+    def _synchronize_if_necessary(self):
+        if not self.synchronize or os.getpid() == self._origin_pid or self._state in TERMINAL_STATES:
+            return
+
+        self._initialize_backend_if_necessary()
+
+        maybe_existing_request = self._status_requests.get(self._uid)
+        if maybe_existing_request is not None:
+            _, status_updated = maybe_existing_request
+            status_updated.wait(MPFuture.HARD_UPDATE_TIMEOUT)
+            return
+
+        # otherwise create a new request for synchronization
+
+        try:
+            status_updated = threading.Event()
+            self._status_requests[self._uid] = (self, status_updated)
```

Comment:
Concurrent calls to `_synchronize_if_necessary` may lead to overwriting an existing `threading.Event` with a new one here. In this case, the older event will never get an update and will break the future after reaching `HARD_UPDATE_TIMEOUT`.